src/Pos/Txp/Toil/Balances/Functions.hs

```diff
@@ -0,0 +1,94 @@
+{-# LANGUAGE ConstraintKinds #-}
+{-# LANGUAGE TypeFamilies    #-}
+
+-- | Functions which work in 'MonadBalances' and are part of Toil logic.
+
+module Pos.Txp.Toil.Balances.Functions
+       ( applyTxsToBalances
+       , rollbackTxsBalances
+       ) where
+
+import           Universum
+
+import qualified Data.HashMap.Strict as HM
+import qualified Data.HashSet        as HS
+import qualified Data.List.NonEmpty  as NE
+import           Formatting          (build, sformat, (%))
+import           System.Wlog         (WithLogger, logInfo)
+
+import           Pos.Core.Coin       (coinToInteger, sumCoins, unsafeAddCoin,
+                                      unsafeIntegerToCoin, unsafeSubCoin)
+import           Pos.Types           (Coin, StakeholderId, mkCoin)
+
+import           Pos.Txp.Core        (Tx (..), TxAux, TxOutAux (..), TxUndo,
+                                      getTxDistribution, txOutStake)
+import           Pos.Txp.Toil.Class  (MonadBalances (..), MonadBalancesRead (..))
+
+type BalancesMode m = (MonadBalances m, WithLogger m)
+
+-- | Apply transactions to balances.
+applyTxsToBalances :: BalancesMode m => [(TxAux, TxUndo)] -> m ()
+applyTxsToBalances txun = do
+    let (txOutPlus, txInMinus) = concatStakes txun
+    recomputeStakes txOutPlus txInMinus
+
+-- | Rollback application of transactions to balances.
+rollbackTxsBalances :: BalancesMode m => [(TxAux, TxUndo)] -> m ()
+rollbackTxsBalances txun = do
+    let (txOutMinus, txInPlus) = concatStakes txun
+    recomputeStakes txInPlus txOutMinus
+
+----------------------------------------------------------------------------
+-- Helpers
+----------------------------------------------------------------------------
+
+-- Compute new stakeholder's stakes by lists of spent and received coins.
+recomputeStakes
+    :: BalancesMode m
+    => [(StakeholderId, Coin)]
+    -> [(StakeholderId, Coin)]
+    -> m ()
+recomputeStakes plusDistr minusDistr = do
+    let needResolve =
+            HS.toList $
+            HS.fromList (map fst plusDistr) `HS.union`
+            HS.fromList (map fst minusDistr)
+    resolvedStakes <- mapM resolve needResolve
+    totalStake <- getTotalStake
+    let positiveDelta = sumCoins (map snd plusDistr)
+    let negativeDelta = sumCoins (map snd minusDistr)
```

Comment:
> reason why you use lets in every line?

I also like `let` on every line better. And aslo because we have this in [style-guide](https://github.com/serokell/serokell-util/blob/867ec23923869aed3ba3367e30ad435b60721060/serokell-style.md#let-expressions)